### PR TITLE
[Integration][AzureDevOps] Live events support for branch kind

### PR DESCRIPTION
**Pull request by dev-habib-nuhu**
### **User description**
# Description

What - Added live event support for Azure DevOps branch entities through webhook processing
Why - Enable real-time synchronization of branch creation, updates, and deletions in Azure DevOps repositories
How - Implemented `BranchWebhookProcessor` that listens to `git.push` events, extracts branch changes from refUpdates, and handles both branch creation/updates and deletions using Azure DevOps zero OID convention

## Type of change

- [x] New feature (non-breaking change which adds functionality)

<h4> All tests should be run against the port production environment(using a testing org). </h4>

### Core testing checklist

- [ ] Integration able to create all default resources from scratch
- [ ] Resync finishes successfully
- [ ] Resync able to create entities
- [ ] Resync able to update entities
- [ ] Resync able to detect and delete entities
- [ ] Scheduled resync able to abort existing resync and start a new one
- [ ] Tested with at least 2 integrations from scratch
- [ ] Tested with Kafka and Polling event listeners
- [ ] Tested deletion of entities that don't pass the selector

### Integration testing checklist

- [ ] Integration able to create all default resources from scratch
- [ ] Completed a full resync from a freshly installed integration and it completed successfully
- [ ] Resync able to create entities
- [ ] Resync able to update entities
- [ ] Resync able to detect and delete entities
- [ ] Resync finishes successfully
- [ ] If new resource kind is added or updated in the integration, add example raw data, mapping and expected result to the `examples` folder in the integration directory.
- [ ] If resource kind is updated, run the integration with the example data and check if the expected result is achieved
- [ ] If new resource kind is added or updated, validate that live-events for that resource are working as expected
- [ ] Docs PR link [here](#)

### Preflight checklist

- [ ] Handled rate limiting
- [ ] Handled pagination
- [ ] Implemented the code in async
- [ ] Support Multi account

## Screenshots

Include screenshots from your environment showing how the resources of the integration will look.

## API Documentation

Provide links to the API documentation used for this integration.

___

### **PR Type**
Enhancement

___

### **Description**
- Implement `BranchWebhookProcessor` for real-time branch event handling

- Process `git.push` webhook events to detect branch creation, updates, deletions

- Extract branch changes from refUpdates using Azure DevOps zero OID convention

- Register processor with webhook handler in main application

- Add comprehensive test coverage for branch webhook processing

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["git.push Event"] --> B["BranchWebhookProcessor"]
  B --> C["Extract refUpdates"]
  C --> D["Filter refs/heads/"]
  D --> E["Detect Changes"]
  E --> F["Updated Branches"]
  E --> G["Deleted Branches"]
  F --> H["WebhookEventRawResults"]
  G --> H
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>branch_webhook_processor.py</strong><dd><code>Branch webhook processor implementation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/azure-devops/azure_devops/webhooks/webhook_processors/branch_webhook_processor.py

<ul><li>Implement <code>BranchWebhookProcessor</code> class extending <br><code>AzureDevOpsBaseWebhookProcessor</code><br> <li> Process <code>git.push</code> events and validate payload structure with repository <br>and refUpdates<br> <li> Extract branch changes from refUpdates, distinguishing updates from <br>deletions using zero OID<br> <li> Filter only branch references (refs/heads/) and exclude tags and <br>remote refs<br> <li> Return structured results with updated and deleted branch data <br>including repository context</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2324/files#diff-fd18e27dadf080ba1b86712ee3504f55ff32de77b5c5f84fae850dd23eed3d0a">+114/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>main.py</strong><dd><code>Register branch webhook processor</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/azure-devops/main.py

<ul><li>Import <code>BranchWebhookProcessor</code> from webhook processors module<br> <li> Register <code>BranchWebhookProcessor</code> with webhook handler at <code>/webhook</code> <br>endpoint</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2324/files#diff-fd3f12e4dacbc3f260d7ce063fbd614eb8e2b2e252b87ab24eef80d2a7cfa401">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_branch_webhook_processor.py</strong><dd><code>Branch webhook processor test suite</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/azure-devops/tests/azure_devops/webhooks/webhook_processors/test_branch_webhook_processor.py

<ul><li>Add comprehensive test suite for <code>BranchWebhookProcessor</code> with 8 test <br>cases<br> <li> Test event filtering for <code>git.push</code> event type and payload validation<br> <li> Test branch update and deletion handling with zero OID detection<br> <li> Test edge cases: missing repository, non-branch refs, malformed data<br> <li> Test static method <code>_extract_branch_changes</code> with various input <br>scenarios</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2324/files#diff-c010e9196722481ab11287a1b177196c45d03e5a37923a355610b1e5de23ffe0">+223/-0</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

**Review by mk-armah — APPROVED** [review]
Add changelog and bump integration version

File: integrations/azure-devops/azure_devops/webhooks/webhook_processors/branch_webhook_processor.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+class BranchWebhookProcessor(AzureDevOpsBaseWebhookProcessor):
+    async def get_matching_kinds(self, event: WebhookEvent) -> list[str]:
+        return [Kind.BRANCH]
+
+    async def should_process_event(self, event: WebhookEvent) -> bool:
+        try:
+            return event.payload.get("eventType") == PushEvents.PUSH
+        except Exception:
+            return False
```
Comment (IdansPort) [inline integrations/azure-devops/azure_devops/webhooks/webhook_processors/branch_webhook_processor.py]: you set get default value for get for example .get("eventType",None) so you wouldn't get exception at all

File: integrations/azure-devops/azure_devops/webhooks/webhook_processors/branch_webhook_processor.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        deleted: List[Dict[str, Any]] = []
+
+        for ref in ref_updates:
+            ref_name = ref.get("name") or ""
+            if not ref_name.startswith("refs/heads/"):
+                continue
+
+            branch_name = ref_name.replace("refs/heads/", "")
+            new_oid = ref.get("newObjectId") or ""
+            old_oid = ref.get("oldObjectId") or ""
```
Comment (IdansPort) [inline integrations/azure-devops/azure_devops/webhooks/webhook_processors/branch_webhook_processor.py]: ref.get("newObjectId","")
